zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/LearnerHandlerMetricsTest.java

```diff
@@ -60,36 +61,40 @@ public void setup() throws IOException {
 
         //adding 5ms artificial delay when sending each packet
         BinaryOutputArchive oa = mock(BinaryOutputArchive.class);
-        doAnswer(new Answer() {
-            @Override
-            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
-                Thread.sleep(5);
-                return  null;
-            }
-        }).when(oa).writeRecord(any(QuorumPacket.class), Matchers.anyString());
+        doAnswer(invocationOnMock -> {Thread.sleep(5); return null;})
+                .when(oa).writeRecord(any(QuorumPacket.class), Matchers.anyString());
+
+        BufferedOutputStream bos = mock(BufferedOutputStream.class);
+        // flush is called when all packets are sent and the queue is empty
+        doAnswer(invocationOnMock -> {allSentLatch.countDown(); return null;}).when(bos).flush();
```

Comment:
The allSentLatch needs to be initialized in setup, otherwise if the other test case in this class, and it's not initialize this latch, it will throw NPE.

Also if those test only used in testMetrics method, but not general, for other tests, then we should move it into testMetrics instead of setup.

Reply:
I added check for null to avoid NPE. I try to separate setup from testing logic.  Initializing the latch I think is part of the testing logic because it might not always be appropriate to set the latch at the beginning of the test. The place where the latch is set and the count the latch is set to depend on what is being tested.   
There is one test method in this test file so far. So I could move all the setup into the test. But again I want to separate setup from testing logic itself. If later new test methods are to be added and they don't share the common setup, then we may need to refactor the setup code--but if they don't share the common setup, they probably shouldn't be in the same test file.